src/Button.tsx

```diff
@@ -0,0 +1,241 @@
+/** @jsx jsx */
+import { jsx } from "@emotion/core";
+import React, { ComponentProps } from "react";
+import * as colors from "./colors";
+import * as CSS from "csstype";
+import { base } from "./typography";
+
+// Types that could use some improvement:
+// * Don't allow `children` and `icon` to be missing
+// * Don't allow `children` when `FAB`
+//
+// I was able to get guarantees to work, but only with very cryptic errors. I
+// decided it'd be best, for the time being, to `throw` if we use things
+// incorrectly.
+interface Props
+  extends React.DetailedHTMLProps<
+    React.ButtonHTMLAttributes<HTMLButtonElement>,
+    HTMLButtonElement
+  > {
+  /**
+   * Treat this button as a floating action button
+   *
+   * You must include an `icon` prop and you must _not_ include a `children`
+   * prop for a floating action button.
+   */
+  fab?: boolean;
+
+  /**
+   * Force the button to be in the active state.
+   *
+   * Useful for Storybook
+   *
+   * @ignore
+   */
+  forceActiveState?: boolean;
+
+  /**
+   * Force the button to be in the focus state.
+   *
+   * Useful for Storybook
+   *
+   * @ignore
+   */
+  forceFocusState?: boolean;
+
+  /**
+   * Force the button to be in the hover state.
+   *
+   * Useful for Storybook
+   *
+   * @ignore
+   */
+  forceHoverState?: boolean;
+
+  /**
+   * Either an icon to show to the left of the button text, or on it's own
+   */
+  icon?: React.ReactElement;
+
+  /**
+   * Size of the button
+   *
+   * Defaults to "default"
+   */
+  size?: "default" | "small" | "large";
+
+  /**
+   * Which variant to display
+   *
+   * The options are as follows:
+   *
+   * - `normal`: A button with a border and a background
+   * - `simple`: No border or background (overrideable).
+   *     - This will require you adding an `:active` style in addition to `:hover`
+   */
+  variant?: "normal" | "simple";
```

Comment:
It previously was a boolean flag; @timbotnik 	suggested we remove the boolean flag and instead use something more descriptive. I'm very happy to use something besides "simple" and "normal". I don't want to bikeshed this though.

Reply:
I found the comment, would you mind writing a bit about the reasoning behind it being more sensible to have a string enum here?  

I agree that `hidden` shouldn't be used since that's appropriating something else, but switching away from a boolean prop makes usage more verbose (`<Button simple/>` vs `<Button variant="basic"/>`), and also makes condition-checks in the component noisier (`props.simple ? ... : ...` vs `props.variant === 'simple' ? ... : ...`). What are these tradeoffs in service of?

Reply:
How about `raised` and `flat`?

Reply:
@trevorblades I like `raised` and `flat`!  The general reasoning behind not using a boolean is so that you can't provide an API that allows contradictions, i.e. `<Button raised flat>`.  But perhaps we'd want to deal with this via PropType checking at run time?  I'm not going to hold my position that firmly, it's not a very big deal to me in the end, so also happy to be the odd man out.

Reply:
I don't think we'd need both props, just one should be enough?   

`<Button/>` would be the raised/normal button
`<Button flat/>` would be flat

Reply:
I will hold strong on not allowing contradicting props; especially where we don't have _any_ dynamic propType checking.

I accidentally used `variant`, and I should have used `feel` as @timbotnik 	originally suggested. I'm totally cool with `raised` and `flat`, as long as they are string props passed somewhere.

I read an article a few weeks ago: [How to Avoid the “Boolean Trap” When Designing React Components](https://spicefactory.co/blog/2019/03/26/how-to-avoid-the-boolean-trap-when-designing-react-components/) and agree that using an enum for the different options feels more correct.

Yes, it's more verbose; I don't think that's a bad thing. It is still readable, the different enum options can be documented in the `interface` so you get helpful type hinting with TS.

Reply:
There is no contradicting prop  

`flat?: boolean` and that's it. If it's true then it's flat, if it's false then it's raised

Reply:
Yea, that makes sense to me. Considering we only have two button variants to represent, we could use a single `flat` boolean. `variant` allows us the ability to add additional variants in the future without changing the API, but we also may not want to worry about future changes/additions at this point.

@timbotnik we're on the same page with regards to the two-boolean thing. 👍  I was suggesting we use `flat` and `raised` as values for the `variant` prop.

Reply:
I think we're coming down to personal preferences and playing devil's advocate now. I'd prefer to leave this as a `feel?: 'raised' | 'flag'`.

Reply:
Specifically regarding the article; Uncle Bob is indeed frequently relevant, but I believe the author of the post is mistaken in using it for his argument here  

"loudly proclaiming that this function does more than one thing" is not a problem that enums solve (if anything, a boolean flag proclaims to do only two things, while an enum proclaims to do an unknown and likely higher number (until you mouse over to see exactly how many)). The problem exists in the scope that a function (/component) has chosen to encapsulate. At the end of the "Flag Arguments" section, he writes "We should have split the function into two: renderForSuite() and renderForSingleTest()." -- the solution to flag arguments in his opinion is to split into separate functions. This is one of those sections that imo might not translate as directly to UI development (unless we choose to split this one component into separate `<SimpleButton/>` and `<Button/>` components)

Another reason why he's so against flag arguments is in the next paragraph of the quote from the article - "[..] the method call `render(true)` is just plain confusing to a poor reader. Mousing over the call and seeing `render(boolean isSuite)` helps a little, but not that much. [..]". boolean props in React don't suffer from the same confusion.

Reply:
I think we've reached a concesnsus that what I've got here is ok and I wanted to leave one more comment after a good night's sleep and after reading reading https://overreacted.io/optimized-for-change/ again.

I want to avoid boolean props because there's a possibility that we'll come up with more button variants than "fab" in the future, so using an optional string prop with the only option being `"fab"` makes those changes backwards compatible. I weighed this decision against "solve today's problems" and decided this solves today's problems well and is optimized for change.